src/ns-support.c

```diff
@@ -0,0 +1,406 @@
+	debug("opening namespace group directory %s", sc_ns_dir);
+	group->dir_fd =
+	    open(sc_ns_dir, O_DIRECTORY | O_PATH | O_CLOEXEC | O_NOFOLLOW);
+	if (group->dir_fd == -1) {
+		die("cannot open directory for namespace group %s", group_name);
+	}
+	char lock_fname[PATH_MAX];
+	must_snprintf(lock_fname, sizeof lock_fname, "%s%s", group_name,
+		      SC_NS_LOCK_FILE);
+	debug("opening lock file for namespace group %s", group_name);
+	group->lock_fd =
+	    openat(group->dir_fd, lock_fname,
+		   O_CREAT | O_RDWR | O_CLOEXEC | O_NOFOLLOW, 0600);
+	if (group->lock_fd == -1) {
+		die("cannot open lock file for namespace group %s", group_name);
+	}
+	group->name = strdup(group_name);
+	if (group->name == NULL) {
+		die("cannot duplicate namespace group name %s", group_name);
+	}
+	return group;
+}
+
+void sc_close_ns_group(struct sc_ns_group *group)
+{
+	debug("releasing resources associated wih namespace group %s",
+	      group->name);
+	close(group->dir_fd);
+	close(group->lock_fd);
+	close(group->event_fd);
+	free(group->name);
+	free(group);
+}
+
+void sc_lock_ns_mutex(struct sc_ns_group *group)
+{
+	debug("acquiring exclusive lock for namespace group %s", group->name);
+	if (flock(group->lock_fd, LOCK_EX) < 0) {
+		die("cannot acquire exclusive lock for namespace group %s",
+		    group->name);
+	}
+	debug("acquired exclusive lock for namespace group %s", group->name);
+}
+
+void sc_unlock_ns_mutex(struct sc_ns_group *group)
+{
+	debug("releasing lock for namespace group %s", group->name);
+	if (flock(group->lock_fd, LOCK_UN) < 0) {
+		die("cannot release lock for namespace group %s", group->name);
+	}
+	debug("released lock for namespace group %s", group->name);
+}
+
+void sc_create_or_join_ns_group(struct sc_ns_group *group)
+{
+	// Open the mount namespace file.
+	char mnt_fname[PATH_MAX];
+	must_snprintf(mnt_fname, sizeof mnt_fname, "%s%s", group->name,
+		      SC_NS_MNT_FILE);
+	int mnt_fd __attribute__ ((cleanup(sc_cleanup_close))) = -1;
+	mnt_fd =
+	    openat(group->dir_fd, mnt_fname,
+		   O_CREAT | O_RDONLY | O_CLOEXEC | O_NOFOLLOW, 0600);
+	if (mnt_fd == -1) {
+		die("cannot open mount namespace file for namespace group %s",
+		    group->name);
+	}
+	// attempt to join an existing group
+	debug
+	    ("attempting to re-associate the mount namespace with the namespace group %s",
+	     group->name);
+	if (setns(mnt_fd, CLONE_NEWNS) == 0) {
+		debug
+		    ("successfully re-associated the mount namespace with the namespace group %s",
+		     group->name);
+		return;
+	}
+	// Anything but EINVAL is an unexpected error.
+	//
+	// EINVAL is simply a sign that the file we've opened is not a valid
+	// namespace file descriptor. One potential case where this can happen is
+	// when another snap-confine tried to initialize the namespace but was
+	// killed before it managed to complete the process.
+	if (errno != EINVAL) {
+		die("cannot re-associate the mount namespace with namespace group %s", group->name);
+	}
```

Comment:
You are right. I plan to change this a little to call `fstatfs()` to determine if the opened file descriptor belongs to `nsfs` or not. This will let us predictably die() when anything goes wrong.